glean-core/src/upload/mod.rs

```diff
@@ -94,99 +127,179 @@ impl PingUploadManager {
+        assert_eq!(upload_manager.get_next_ping(), PingsQueueStatus::Empty);
+    }
+
+    #[test]
+    fn test_fills_up_queue_successfully_from_disk() {
+        let (mut glean, dir) = new_glean(None);
+
+        // Register a ping for testing
+        let ping_type = PingType::new("test", true, true);
+        glean.register_ping_type(&ping_type);
 
         // Submit the ping multiple times
         let n = 10;
         for _ in 0..n {
-            upload_manager.enqueue_ping(uuid, url, body);
+            glean.submit_ping(&ping_type).unwrap();
         }
 
-        // Clear the queue
-        glean.upload_manager.clear_ping_queue();
+        // Create a new upload_manager
+        let upload_manager = PingUploadManager::new();
+
+        // Get the data path from the glean instance as &str
+        let data_path = dir.path().to_str().unwrap();
+
+        // Attempt at processing the directory
+        let handle = upload_manager.process_pending_pings(&data_path);
+        // Join the thread to ensure we are done processing before continuing
+        handle.join().unwrap();
 
-        // Verify there really isn't any ping in the queue.
-        assert!(glean.upload_manager.get_next_ping().is_none());
+        // Verify the requests were properly enqueued
+        for _ in 0..n {
+            match upload_manager.get_next_ping() {
+                PingsQueueStatus::Next(_) => {}
+                _ => panic!("Expected upload manager to return the next request!"),
+            }
+        }
+
+        // Verify that after all requests are returned, none are left
+        assert_eq!(upload_manager.get_next_ping(), PingsQueueStatus::Empty);
+    }
+
+    #[test]
+    fn test_processing_pending_pings_flag_works() {
+        let upload_manager = PingUploadManager::new();
+
+        // With nothing queued and no processing going on, flag should be `false`...
+        assert_eq!(upload_manager.is_processing_pending_pings(), false);
+        // ...and `get_next_ping` should return `Empty`.
+        assert_eq!(upload_manager.get_next_ping(), PingsQueueStatus::Empty);
+
+        // Create a new glean to help on submitting pings
+        let (mut glean, dir) = new_glean(None);
+
+        // Register a ping for testing
+        let ping_type = PingType::new("test", true, true);
+        glean.register_ping_type(&ping_type);
+
+        // Submit the ping
+        glean.submit_ping(&ping_type).unwrap();
+
+        // Get the data path from the glean instance as &str
+        let data_path = dir.path().to_str().unwrap();
+
+        // Attempt at processing the directory
+        let handle = upload_manager.process_pending_pings(&data_path);
+
+        // With the processing ongoing, flag should be `true`
+        assert_eq!(upload_manager.is_processing_pending_pings(), true);
+        // ...and `get_next_ping` should return `Processing`
+        assert_eq!(upload_manager.get_next_ping(), PingsQueueStatus::Processing);
+
```

Comment:
I am not sure the two above assertions are not going to create intermittents, because I don't have such fine grained control over order of execution on the spawned thread. @badboy what do you think?

Reply:
True. I think the flag is merely an implementation detail.

If we go with what I said before that we don't toggle back and forth between the states what we would want to test here is:

* When starting, we have not processed the directory yet and don't get a ping out
* We start processing and wait for it to finish.
* After the async processing is done, we get a ping out

Reply:
Yes, I agree with you. And that would mean just removing this test altogether, because what you are describing is what the test just before this one does.